src/geodesy.jl

```diff
@@ -0,0 +1,433 @@
+import Base.isgreater
+import DataStructures.Queue, DataStructures.Deque, DataStructures.enqueue!, DataStructures.dequeue!, DataStructures.dequeue_pair!
+
+_build_connectivity(r::Int, n) = trues(ntuple(_->r, n))
+function _build_connectivity(r::NTuple{N,Int}, n) where N
+    N <= n || throw(ArgumentError("radius length $(length(r)) must be less or equal to $n"))
+    sz = (r..., ntuple(_->1, n-N)...)
+    return trues(sz)
+end
+
+"""
+    hmaximum(img, h; r=3)
+
+For grayscale image `img`, the h-maximum transformation suppresses all regional maxima whose
+depth is below a given threshold level `h`. The regional size `r` can be either integer or
+tuple of integer.
+
+```julia
+using ImageMorphology, TestImages
+
+img = Gray.(testimage("blob"))
+img_h = hmaximum(img, 0.4)
+
+# suppress the regional maxima along the second dimension
+img_h = hmaximum(img, 0.4; r=(1, 3))
+```
+
+See also [`hminimum`](@ref).
+"""
+function hmaximum(img::AbstractArray, h; r=3)
+    # TODO(johnnychen94): support in-place version
+    connectivity = _build_connectivity(r, ndims(img))
+    tmp = mappedarray(x -> x - h, img)
+    return underbuild(tmp, img, connectivity)
+end
+
+"""
+    hminimum(img, h)
+
+For grayscale image `img`, the h-minimum transformation suppresses all regional maxima whose
+depth is above a given threshold level `h`.
+
+```julia
+using ImageMorphology, TestImages
+
+img = Gray.(testimage("blob"))
+img_h = hminimum(img, 0.4)
+
+# suppress the regional minimum along the second dimension
+img_h = hminimum(img, 0.4; r=(1, 3))
+```
+
+See also [`hminimum`](@ref).
+"""
+function hminimum(img::AbstractArray, h; r=3)
+    # TODO(johnnychen94): support in-place version
+    connectivity = _build_connectivity(r, ndims(img))
+    tmp = mappedarray(x -> x + h, img)
+    return overbuild(tmp, img, connectivity)
+end
+
+"""
+    underbuild(marker, mask, connectivity=[]) -> Array
+
+Morphological reconstruction by dilatation according to connection type
+
+# Details
+This operation performs dilations of marker restricted by the provided mask until idempotence This means that the output image
+will be included between mask and min(marker,mask).
+# Arguments
+- `marker::AbstractArray{T, N}`: where {T<:ImageCore.NumberLike , N} the ``N``-dimensional input image
+- `mask::AbstractArray{T, N}`: where {T<:ImageCore.NumberLike , N} the ``N``-dimensional input image
+- `connectivity::AbstractArray{Bool}`: the neighborhood connectivity.
+# Returns
+An array of the same type and shape as the `marker`.
+# See also
+[`overbuild`](@ref)
+# References
+  Morphological Grayscale Reconstruction in Image Analysis: Applications and Efficient Algorithms IEEE Trans Image Process. 1993;2(2)
+"""
+function underbuild(marker::AbstractArray, mask::AbstractArray, connectivity::AbstractArray{Bool})
+    N = ndims(marker)
+    # check_image_consistency(mask, marker)
+
+    all(in((1, 3)), size(connectivity)) || throw(ArgumentError("connectivity must have size 1 or 3 in each dimension"))
```

Comment:
@ThomasRetornaz a quick question, is there any reason we limit ourselves to `1` and `3` here? If not, then I'll add a TODO note here.

Reply:
Basically reconstruction algorithm make sens if we use something related to an unit "circle/ball" structuring element
We could use symetric segement in rare cases
May the test here is not correct

Reply:
Nope sorry only the "unit" circle aka r=1 make sens. But i never think about much, the algorithm will be ok (aka not crash)
But i think theorically and pratically we gain nothing supporting r >1

I will review twice your tests

Reply:
> Nope sorry only the "unit" circle aka r=1 make sens. But i never think about much, the algorithm will be ok (aka not crash)
But i think theorically and pratically we gain nothing supporting r >1

In this case, I think `dims` keyword is more meaningful:

```julia
hmaxima(img, h) == hmaxima(img, h; dims=(1, 2))
hmaxima(img, h; dims=(1, )) # a fat region along dimension 1
```

Reply:
Yes but no :) because in 2D we could have C4 and C8 connectivity and in 3D at least C26 C6 connectivity
dims=1 limits to C8/C26 
But anyway in this case its not a drama. Just if i made an erode/dilate with a certain connectivity i have to keep the same connectivity for reconstruction i think, but even me i don't pay to much attention
For the moment erode/dilate function supports dims approach but its too restrictive
Erode/DIlate must support at least segment and no symetric SE -> an other incoming (i hope) PR

Note that in theory the problem is solved by using an hexagonal grid and a C6 connectivity in 2D